packages/app-mobile/components/NoteEditor/CodeMirror/RegionSpec.ts

```diff
@@ -0,0 +1,153 @@
+/**
+ * Describes an inline region's formatting. For example, an italicized region.
+ */
+
+import { Text as DocumentText, SelectionRange } from '@codemirror/state';
+
+export default class RegionSpec {
+	// Patterns to use when creating an instance of the tag.
+	// E.g.
+	//   templateStart=**
+	//   templateStop=**
+	//  would be used to create
+	//   **content here**
+	public templateStart: string;
+	public templateStop: string;
+
+	// Size of the buffer used to check for a match.
+	// Particularly useful for [startExp], [stopExp].
+	// Only used when the selection is empty.
+	protected startBuffSize: number;
+	protected stopBuffSize: number;
+
+	// Regular expressions for matching possible starting/stopping
+	// regions.
+	protected startExp?: RegExp;
+	protected stopExp?: RegExp;
+
+	public constructor({
+		templateStart, templateStop,
+		startExp, stopExp,
+		startBuffSize, stopBuffSize,
+	}: {
+			templateStart: string; templateStop: string;
+			startExp?: RegExp; stopExp?: RegExp;
+			startBuffSize?: number; stopBuffSize?: number;
+	}) {
+		this.templateStart = templateStart;
+		this.templateStop = templateStop;
+		this.startExp = startExp;
+		this.stopExp = stopExp;
+		this.startBuffSize = startBuffSize ?? this.templateStart.length;
+		this.stopBuffSize = stopBuffSize ?? this.templateStop.length;
+	}
+
+
+	/**
+	 * Returns the length of a match for this in [searchText], matching either
+	 * [template] or [regex], starting at [startIndex] or ending at [endIndex].
+	 * Note that the [regex], if provided, must have the global flag enabled.
+	 */
+	private matchLen({
+		searchText, template, startIndex, endIndex, regex,
+	}: {
+		searchText: string; template: string;
+		startIndex: number; endIndex: number; regex?: RegExp;
+	}): number {
+		let matchLength, matchIndex;
+
+		// Returns true if [idx] is in the right place (the match is at
+		// the end of the string or the beginning based on startIndex/endIndex).
+		const indexSatisfies = (idx: number, len: number): boolean => {
+			return (startIndex == -1 || idx == startIndex)
+				&& (endIndex == -1 || idx == endIndex - len);
+		};
+
+		if (regex) {
+			// Enforce 'g' flag.
+			if (regex.flags.indexOf('g') == -1) {
+				throw new Error('Regular expressions used by RegionSpec must have the global flag!');
+			}
+
+			// Search from the beginning.
+			regex!.lastIndex = 0;
```

Comment:
What is this "regex!" thing? Can't the code be written without this?

Reply:
The `!` is the [non-null assertion operator](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html#non-null-assertion-operator).

We're not currently using [`--strictNullChecks`](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html#--strictnullchecks), but if we do, we'll need these assertions.

(I've started being lazy — I'm not using `foo!.bar` syntax in newer code)

Reply:
But you have a `if (regex)` before that code so you already know it's not null.

I'd rather we don't add such assertions. We should normally be confident that input/output are correct, via test units and type checking, and we can have explicit null/undefined check if we know such value can happen in normal operations.